File: Semester1/HW_8/10_11_Task3/10_11_Task3/stack.cpp
#include "stack.h"

struct StackElement
{
	int value;
	StackElement * next;
};

struct Stack
{
	StackElement * head = nullptr;
};

Stack * createStack()
{
	Stack * stack = new Stack;
	return stack;
}

void stackPush(Stack *stack, int value)
{
	StackElement *newElement = new StackElement;
	newElement->value = value;
	newElement->next = stack->head;
	stack->head = newElement;
}

int stackPop(Stack * stack)
{
	int value = stack->head->value;
	StackElement * temp = stack->head->next;
	delete stack->head;
	stack->head = temp;
	return value;
}

void deleteStack(Stack * stack)
{
	while (stack->head != nullptr)
	{
		StackElement *temp = stack->head->next;
		delete stack->head;
		stack->head = temp;
	}

	delete stack;
}

File: Semester2/HW_2/Task1/StackClass/StackClass/StackClass/Stack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace StackClass
{
    public class Stack
    {
        private class StackElement
        {
            public int Value { get; set; }
            public StackElement Next { get; set; }
        }

        /// <summary>
        /// Add element to stack.
        /// </summary>
        /// <param name="value"></param>
        public void Push(int value)
        {
            var newElement = new StackElement()
            {
                Next = head,
                Value = value
            };
            head = newElement;
        }

        /// <summary>
        /// Remove element from stack, if it possible.
        /// </summary>
        /// <returns></returns>
        public int Pop()
        {
            if (head == null)
            {
                throw new ExtractFromEmptyStackException("You tried to extract element from empty stack");
            }
            var temp = head.Value;
            head = head.Next;
            return temp;
        }

        /// <summary>
        /// Check if stack is empty.
        /// </summary>
        /// <returns></returns>
        public bool IsEmpty()
        {
            return head == null;
        }

        private StackElement head = null;
    }
}